lib/grandpa/message_handler.go

```diff
@@ -419,36 +451,40 @@ func (h *MessageHandler) verifyPreVoteJustification(msg *CatchUpResponse) (commo
 }
 
 func (h *MessageHandler) verifyPreCommitJustification(msg *CatchUpResponse) error {
-	for _, pcj := range msg.PreCommitJustification {
-		err := verifyBlockHashAgainstBlockNumber(h.blockState, pcj.Vote.Hash, uint(pcj.Vote.Number))
-		if err != nil {
-			if errors.Is(err, chaindb.ErrKeyNotFound) {
-				h.grandpa.tracker.addCatchUpResponse(msg)
-				logger.Infof("we might not have synced to the given block %s yet: %s", pcj.Vote.Hash, err)
-				continue
-			}
-			return err
-		}
-	}
-
 	auths := make([]AuthData, len(msg.PreCommitJustification))
 	for i, pcj := range msg.PreCommitJustification {
 		auths[i] = AuthData{AuthorityID: pcj.AuthorityID}
 	}
 
+	err := verifyIfDescendantOfHighestFinalisedBlock(h.blockState, msg.Hash)
+	if err != nil {
+		return err
+	}
+
 	eqvVoters := getEquivocatoryVoters(auths)
 
 	// verify pre-commit justification
 	var count uint64
 	for idx := range msg.PreCommitJustification {
 		just := &msg.PreCommitJustification[idx]
 
-		if _, ok := eqvVoters[just.AuthorityID]; ok {
-			continue
+		err = verifyBlockHashAgainstBlockNumber(h.blockState, just.Vote.Hash, uint(just.Vote.Number))
+		if err != nil {
+			if errors.Is(err, chaindb.ErrKeyNotFound) {
+				h.grandpa.tracker.addCatchUpResponse(msg)
+				logger.Infof("we might not have synced to the given block %s yet: %s", just.Vote.Hash, err)
```

Comment:
Less identantion:
```suggestion
		if errors.Is(err, chaindb.ErrKeyNotFound) {
				h.grandpa.tracker.addCatchUpResponse(msg)
				logger.Infof("we might not have synced to the given block %s yet: %s", just.Vote.Hash, err)
				continue
	        } else if err != nil {
				return err
	        }   
```

Reply:
:smile: I agree, although Tim previously suggested to have the `errors.Is` inside the `if err != nil` block... I would just keep it as it is I think :laughing:

Reply:
I do not find this easier to read.  If there's an error handle it within the block.  There shouldn't be multiple `if err != nil` lines.

Reply:
I do not find nested if's easy to read and usually what I suggest for long errors case is to use `switch err`. eg:
```go
switch err {
case ErrSome:
case ErrAnother:
case nil:
}
```

and for simple error checks using the:

```go
if errors.Is(err, ErrToCheck) {
     handle here...
} else err != nil {
     handle the other errors here
}
```

Reply:
resolving, this does not seem to need any action